# Работа с данными 

Предположим в июле провели специальную акцию, направленную на улучшение ретеншена. Новым юзерам слали письмо с купоном. Сработала ли акция?

https://drive.google.com/file/d/18CZKLXpbZfR44SunmzSuFvc0kI9U8n_K/view?usp=sharing

сделать можно на питоне в юпитер или экселе или в гугл шите, даже можете в sql сделать, как удобнее.
обратно нужно дать файл с формулами или кодом и описание того, как получился результат поэтапно.


In [170]:
import numpy as np
import pandas as pd

# импорт данных
df = pd.read_csv('analyst_test_transactions_db.csv', delimiter=',')
df.head()

user_id  order_id    reg_date transaction_date  revenue
0    57562   94545.0  2013-12-22       2015-07-16     1088
1    40047   81824.0  2013-11-24       2015-01-19     1027
2    48153  144851.0  2015-07-03       2015-10-29      601
3    54591  116111.0  2013-07-12       2015-04-30      881
4    39789  134943.0  2013-05-25       2015-07-21      984

In [2]:
df = df.sort_values(['reg_date'], ascending=[False])
df.head(5)

user_id  order_id    reg_date transaction_date  revenue
37842    53201  115680.0  2015-11-20       2015-11-20      749
51574    41823  145416.0  2015-11-20       2015-11-20     1118
3091     57593   46617.0  2015-11-20       2015-11-20      713
57677    56308  -16479.0  2015-11-20       2015-11-20      940
75303    38278  -21334.0  2015-11-20       2015-11-20     1299

In [3]:
# вывод информации по данным
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 76857 entries, 37842 to 69591
Data columns (total 5 columns):
user_id             76857 non-null int64
order_id            76857 non-null float64
reg_date            76857 non-null object
transaction_date    76857 non-null object
revenue             76857 non-null int64
dtypes: float64(1), int64(2), object(2)
memory usage: 3.5+ MB


In [4]:
# вывод количества уникальных значений для анализа.
df.nunique() 

user_id             48214
order_id            76857
reg_date             1363
transaction_date      324
revenue              1630
dtype: int64

In [2]:
# вывод количества уникальных клиентов
df.user_id.nunique()

48214

In [3]:
# вывод количества уникальных клиентов с датой регистрации до июля
dk=df[df['reg_date'] < '2015-07-01']
print(dk.user_id.nunique())
dk.head()

42872


user_id  order_id    reg_date transaction_date  revenue
0    57562   94545.0  2013-12-22       2015-07-16     1088
1    40047   81824.0  2013-11-24       2015-01-19     1027
3    54591  116111.0  2013-07-12       2015-04-30      881
4    39789  134943.0  2013-05-25       2015-07-21      984
6    50977  136343.0  2012-06-27       2015-03-13     1028

In [4]:
# вывод количества уникальных клиентов с датой регистрации после июля
dg=df[df['reg_date'] >= '2015-07-01']
print(dg.user_id.nunique())
dg = dg.sort_values(['reg_date'])
dg.head()
# в результате увидели несостыковку в данных, а именно 42872 + 12105 != 48214

12105


user_id  order_id    reg_date transaction_date  revenue
27086    59846  112803.0  2015-07-01       2015-09-21     1080
2713     39727   56227.0  2015-07-01       2015-07-23     1089
4926     96264   62460.0  2015-07-01       2015-07-23     1228
3446     24304  171580.0  2015-07-01       2015-10-31      736
25462    27833   97175.0  2015-07-01       2015-11-16     1288

In [59]:
# поиск самого часто встречаемого id клиента
dd = df.groupby(['user_id']).size().reset_index(name='count')
dd=dd[dd['count'] > 8]
dd.head(20)

user_id  count
21454    46832      9

In [60]:
# вывод для анализа даты регистрации
ds=df[df['user_id'] == 46832]
ds=ds.sort_values(['reg_date'])
ds.head(10)

user_id  order_id    reg_date transaction_date  revenue
14143    46832  116348.0  2013-10-25       2015-06-16     1325
65498    46832  119859.0  2014-01-13       2015-02-19     1093
11357    46832  168631.0  2014-01-27       2015-05-06      906
55924    46832  119140.0  2014-04-30       2015-07-29     1251
18583    46832   54590.0  2014-09-12       2015-07-31     1254
633      46832  120641.0  2014-10-11       2015-11-07      535
27617    46832  106417.0  2014-10-23       2015-03-03     1032
52091    46832  110713.0  2015-02-15       2015-11-12     1344
20814    46832  -10657.0  2015-04-17       2015-08-06      656

Мы видим, что с одним и тем же user_id дата регистрации указана разная, надо исправить reg_date для одинаковых user_id.

In [204]:
dd = df.groupby(['user_id']).size().reset_index(name='count')
new=pd.merge(df, dd, how ='inner', on ='user_id')
new.head(10)

user_id  order_id    reg_date transaction_date  revenue  count
0    57562   94545.0  2013-12-22       2015-07-16     1088      2
1    57562   56350.0  2013-07-07       2015-11-01     1182      2
2    40047   81824.0  2013-11-24       2015-01-19     1027      3
3    40047   92251.0  2014-10-05       2015-09-02     1346      3
4    40047   84376.0  2015-02-10       2015-02-25      852      3
5    48153  144851.0  2015-07-03       2015-10-29      601      1
6    54591  116111.0  2013-07-12       2015-04-30      881      4
7    54591   78414.0  2015-01-19       2015-06-27      851      4
8    54591  115354.0  2015-01-17       2015-09-13      908      4
9    54591  152028.0  2014-12-03       2015-04-08     1110      4

In [205]:
bd=new
bb=new

Для исправления даты регистрации запускаем следующий шаг:

In [778]:
for i in range(1000):
    bd=bb[bb['count'] > 1]
    bd=bd.sort_values(['reg_date'])
    bd=bd.sort_values(['user_id'])
    idd=min(bd['user_id'])
    bd=bd[bd['user_id'] ==idd]
    bd=bd.sort_values(['reg_date'])
    mindate=min(bd['reg_date'])
    bb['reg_date'] = np.where(bb['user_id'] == idd, mindate, bb['reg_date'])
    bb['count'] = np.where(bb['user_id'] == idd, 1, bb['count'])

ValueError: min() arg is an empty sequence

In [783]:
# проверка того, что больше не осталось данных с неверной датой регистрации по count
bb.nunique() 

user_id             48214
order_id            76857
reg_date             1363
transaction_date      324
revenue              1630
count                   1
dtype: int64

In [793]:
bb=bb.drop(['count'], axis = 1)

In [794]:
# сохраняем исправленный набор, выложен в репозитории
bb.to_csv('new_analyst_test1.csv')

Retention Rate -это показатель того, сколько людей остаются вашими клиентами на протяжении определенного количества времени. Этот коэффициент можно измерить например за месяц, квартал или год. Он отображает, насколько высок уровень лояльности клиентов к вашему бизнесу.

Формула для вычисления данного показателя выглядит так:

 **Retention = ( CE - CN ) \ CS * 100**

**CE** - это количество клиентов на конец периода.

**CN** - это количество новых клиентов, приобретенных за период.

**CS** - это количество клиентов на начало периода.

In [804]:
# вывод количества клиентов до акции
d1=bb[bb['reg_date'] < '2015-07-01']
print(d1.user_id.nunique())
CS=d1.user_id.nunique()

42872


In [805]:
# вывод колиства клиентов, которые ушли. 
# Исходила из той логики, что эти пользователи были зарегестрированы до акции и после июля не выволняли транзакции
d2=d1[d1['transaction_date'] < '2015-07-01']
print(d2.user_id.nunique())
x=d2.user_id.nunique()

27275


In [806]:
# вывод количества новых клиентов
d3=bb[bb['reg_date'] >= '2015-07-01']
print(d3.user_id.nunique())
CN=d3.user_id.nunique()

5342


In [807]:
# вывод количества клиентов после акции
CE=CS+(CN-x)
print(CE)

20939


In [810]:
Retention=(CE-CN)/CS*100
print(Retention)

36.3803881321142


Исходя из того, что Retention=36.38% можно сделать вывод о том, что акция прошла неудачно.

*Но допускаю возможность, что я неправильно рассчитала количество ушедших клиентов*